### PR TITLE
Release G.5.1 / H.2.1

### LINKED ISSUE #320: Feature request: Cliff Sensor Intensity Values
### Discussed in https://github.com/iRobotEducation/create3_docs/discussions/319

<div type='discussions-op-text'>

<sup>Originally posted by **Kola50011** January 24, 2023</sup>
Hey! 
I am working on a project, where we would like to detect a black line on the floor using the 4 cliff sensors. Sadly, this information not published by the ROS API. Is there any way to access such information, other than the _/hazard_detection_ topic? We  would greatly appreciate something similar to the _/ir_intensity_ topic.</div>

### LINKED ISSUE #333: Improve configurability of NTP sync sources
As an end user it would be helpful to edit the chrony.conf file on the create3, using a similar process as the "Application ROS2 Parameter File". So that we can configure the create3 to directly source from a local ntp time server.

_Originally posted by @Jayden-F in https://github.com/iRobotEducation/create3_docs/issues/327#issuecomment-1428869938_

### LINKED ISSUE #400: Web Server: Address of Fast DDS discovery server does NOT accept special characters
Hi,

I found an issue while attempting to communicate 2 create3's (with a RPi4) with another PC. 

In order to connect to both robots from another PC, the discovery servers must be run on different IDs. 

This means that the nodes running from one of the create3 must have the variable `ROS_DISCOVERY_SERVER=;192.168.186.3:11811` (note the preceding `;`). Refer to DDS docs [here](https://fast-dds.docs.eprosima.com/en/latest/fastdds/env_vars/env_vars.html#env-vars-ros-discovery-server) and to my discussion with @EduPonz [here](https://github.com/eProsima/Fast-DDS/discussions/3560#discussioncomment-6077189).

However, when I try to add this in the web server App config, I get the message "Please set a valid ipv4 address and port for the discovery server."
envvar

Can this be changed so special characters are allowed in that field?

Thank you

### LINKED ISSUE #407: Enhancement: Allow users to set XML config files for DDS
I think it would be great to be able to set XML configuration files for the create3, both for Fast-DDS and CycloneDDS. This allows for more flexibility while setting up multi-robot networking, like specifying domain participants, UDPv4/v6, etc:

https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/xml_configuration.html

### LINKED ISSUE #410: Create 3 Motor stall event sends device 0 command 29
for some reason the create3 sends the stall event packet as device 0 instead of device 1

### PR DESCRIPTION
**Pull request by shamlian**
### ROS 2
* Publications
    * Add `/cliff_intensity` topic (fixes #320)
    * Add tf frame for the IMU (called `imu`), to which the `/imu` topic is now connected
* Middleware
    * Allow the use of `;` characters in Fast-DDS Discovery Server configuration (fixes #400)
    * Beta: Allow user to provide XML configuration file for their preferred middleware (fixes #407)
### Core Robot
* Webserver / OS
    * Beta: Allow user full control to edit ntp.conf (fixes #333)
    * Remove erroneous log messages about the robot's lost Wi-Fi connection when it had never been connected in the first place

### iRobot® Education Bluetooth protocol mode
* Fix device number for stall event to match protocol document (fixes #410)

### CODE REVIEW

**Review by justinIRBT — APPROVED** [review]

File: docs/api/hazards.md [hunk 1]
```diff
@@ -28,16 +28,29 @@ By "kidnap" we denote the action of manually lifting the robot and, eventually,
 The Create® 3 robot combines together different sensors data in order to determine when it's being kidnapped.
 A boolean status will be periodically published on the `kidnap_status` topic.
 
+### Reflective Infrared Sensors
 
-### The `ir_intensity` topic
+The Create 3 robot is equipped with 11 reflective IR sensors to detect hazards.
 
-Besides the aforementioned sensors, the Create® 3 robot is also equipped with [7 sets of IR emitters and receivers](/hw/mechanical/#ir-proximity-sensors) and it can use them to detect objects at close range.
+#### The `ir_intensity` topic
```
Comment (brianabouchard) [inline docs/api/hazards.md:35]: This is a different level header than the ones that describe the hazard_detection and kidnap_status topics above as a result of having the Reflective Infrared Sensors header in line 31. Perhaps change the header for the hazard and kidnap to match?

File: docs/api/hazards.md [hunk 2]
```diff
@@ -28,16 +28,29 @@ By "kidnap" we denote the action of manually lifting the robot and, eventually,
...
+
+[7 sets of IR emitters and receivers](/hw/mechanical/#ir-proximity-sensors) are available in the front bumper to detect objects at close range.
```
Comment (brianabouchard) [inline docs/api/hazards.md:37]: Do we publish the actual range of those sensors anywhere? I'm genuinely curious to know.

File: docs/api/hazards.md [hunk 3]
```diff
@@ -28,16 +28,29 @@ By "kidnap" we denote the action of manually lifting the robot and, eventually,
...
 The higher the intensity value is, the closer an obstacle is to the robot.
 The `header` field of each indivdual detection will provide all the information required to localize it.
 In particular the timestamp will denote when the robot detected the hazard and the frame id will denote the location of the sensor that performed the detection.
 
+#### The `cliff_intensity` topic
+
+[4 sets of IR emitters and receivers](/hw/mechanical/#bottom-view) are available in on the bottom of the robot just behind its front bumper to detect presence or absence of the floor.
+
+The Create® 3 robot will periodically publish on the `cliff_intensity` topic the raw intensity readings obtained from these sensors.
+The message will be a vector where each element corresponds to a different sensor.
```
Comment (brianabouchard) [inline docs/api/hazards.md:50]: Is it worth providing sample data to show what it should look like? In this case the .msg doesn't give much information around expected values. Same question above for the ir_intensity.

File: docs/releases/g_5_1.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+See below for details.
+
+## Changelog (from G.4.5)
+### ROS 2
+* Publications
+    * Add `/cliff_intensity` topic [(#320)](https://github.com/iRobotEducation/create3_docs/issues/320)
+    * Add tf frame for the IMU (called `imu`), to which the `/imu` topic is now connected
+* Middleware
+    * Allow the use of `;` characters in Fast-DDS Discovery Server configuration [(#400)](https://github.com/iRobotEducation/create3_docs/issues/400)
+    * Beta: Allow user to [provide XML configuration file](../../webserver/rmw-profile-override/) for their preferred middleware [(#407)](https://github.com/iRobotEducation/create3_docs/issues/407)
```
Comment (brianabouchard) [inline docs/releases/g_5_1.md:22]: Missing "an" before XML

File: docs/releases/h_2_1.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+See below for details.
+
+## Changelog (from H.1.2)
+### ROS 2
+* Publications
+    * Add `/cliff_intensity` topic [(#320)](https://github.com/iRobotEducation/create3_docs/issues/320)
+    * Add tf frame for the IMU (called `imu`), to which the `/imu` topic is now connected
+* Middleware
+    * Allow the use of `;` characters in Fast-DDS Discovery Server configuration [(#400)](https://github.com/iRobotEducation/create3_docs/issues/400)
+    * Beta: Allow user to [provide XML configuration file](../../webserver/rmw-profile-override/) for their preferred middleware [(#407)](https://github.com/iRobotEducation/create3_docs/issues/407)
```
Comment (brianabouchard) [inline docs/releases/h_2_1.md:25]: Missing "an" before XML

File: docs/webserver/rmw-profile-override.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+!!!warning
+    Please note that this is a beta feature, and as such is not supported by the customer service team.
+    Please exercise caution, as improper use of beta features may result in an inoperable robot.
+
+![Picture of Override RMW Profile page](data/rmw-profile-override.png)
+
+This page allows the user to replace the robot's middleware XML profile file with one of their choosing. Leave this box blank in order to use the robot's default profile file. After pressing "save," the application must be [restarted](../webserver/application/#restart-application).
+
+## Example for Fast DDS
```
Comment (brianabouchard) [inline docs/webserver/rmw-profile-override.md:12]: I wonder if a brief sentence explaining what this custom XML profile does in comparison to the default might be helpful? Perhaps not the level of detail you would want for a beta release, but I'm thinking longer term.

File: docs/webserver/rmw-profile-override.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# iRobot® Create® 3 Webserver - Override RMW Profile
+The Override RMW Profile page of the Create® 3 webserver allows the user to override the robot's inbuilt RMW profiles for its middleware.
+
+!!!warning
+    Please note that this is a beta feature, and as such is not supported by the customer service team.
+    Please exercise caution, as improper use of beta features may result in an inoperable robot.
+
+![Picture of Override RMW Profile page](data/rmw-profile-override.png)
+
+This page allows the user to replace the robot's middleware XML profile file with one of their choosing. Leave this box blank in order to use the robot's default profile file. After pressing "save," the application must be [restarted](../webserver/application/#restart-application).
```
Comment (brianabouchard) [inline docs/webserver/rmw-profile-override.md:10]: This link isn't working for me for some reason when I built using mkdocs. I don't get an error in the build, but when I am navigating the site I get a 404.

File: docs/webserver/set-datetime.md [hunk 8]
```diff
@@ -1,10 +1,11 @@
 # iRobot® Create® 3 Webserver - Set Date and Time
 The Set Date and Time page of the Create® 3 webserver allows the user to directy change the robot's date and time.
```
Comment (brianabouchard) [inline docs/webserver/set-datetime.md:2]: Will this disable time sync even if it is connected to wifi?

File: docs/webserver/set-wired-subnet.md [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
+!!!warning
+    Please note that this is a beta feature, and as such is not supported by the customer service team.
+    Please exercise caution, as improper use of beta features may result in an inoperable robot.
+
 
-Please note that this is a beta feature, and as such is not supported by the customer service team.
-Please exercise caution, as improper use of beta features may result in an inoperable robot.
+![Picture of set wired subnet page](data/set-wired-subnet.png)
 
 This page allows the user to change the robot's /24 subnet within the class C private address area (192.168.nnn.0/24). After pressing "save," the robot must have its power cycled in order for the change to be made.
```
Comment (brianabouchard) [inline docs/webserver/set-wired-subnet.md:11]: Missing return between lines :)

**Review by brianabouchard — COMMENTED** [review]
A few misc comments, nothing major except that I can't get that restart link to work.

File: docs/webserver/set-datetime.md (hunk 8 above)
Comment (shamlian) [reply docs/webserver/set-datetime.md:2]: In the default configuration, effectively, yes, if the time jumps "enough" for the NTP daemon to assume that the times it's receiving from its sync sources are incorrect, which is why the NTP daemon has to be restarted after it is reconfigured.

File: docs/webserver/set-wired-subnet.md (hunk 9 above)
Comment (shamlian) [reply docs/webserver/set-wired-subnet.md:11]: Ooh, you got me.

File: docs/webserver/rmw-profile-override.md (hunk 7 above)
Comment (shamlian) [reply docs/webserver/rmw-profile-override.md:10]: I made an error, I will fix it.

File: docs/webserver/rmw-profile-override.md (hunk 6 above)
Comment (shamlian) [reply docs/webserver/rmw-profile-override.md:12]: I like where you're going, but up with a succinct way to describe all of the things that can be tweaked in these profile files is making the space between my hears hum. Happy to take any suggestions; I'll think about it.

File: docs/api/hazards.md (hunk 1 above)
Comment (shamlian) [reply docs/api/hazards.md:35]: I thought about this, but I didn't want to break the hierarchy. Let me play with it.

Comment (shamlian) [reply docs/api/hazards.md:35]: Looking at it, I think the extra heading I added is unnecessary; I'm just going to get rid of it and put everything on the same level.

File: docs/webserver/rmw-profile-override.md (hunk 7 above)
Comment (alsora) [inline docs/webserver/rmw-profile-override.md:10]: We should include somewhere the default XMLs

File: docs/api/hazards.md (hunk 2 above)
Comment (shamlian) [reply docs/api/hazards.md:37]: I don't know that we do. I believe the IR intensity is [0, 4096) and the cliff intensity is (-4095, 4096), but I need to verify this. Probably we should be clamping the cliff intensity at zero.

File: docs/webserver/rmw-profile-override.md (hunk 6 above)
Comment (alsora) [reply docs/webserver/rmw-profile-override.md:12]: We should link to Fast-DDS and cycloneDDS docs 
 - https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/xml_configuration.html
 - https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/manual/options.md

Comment (brianabouchard) [reply docs/webserver/rmw-profile-override.md:12]: Ok, totally fair. Perhaps a link to the Fast DDS and Cyclone docs that go over that instead?

File: docs/webserver/rmw-profile-override.md (hunk 7 above)
Comment (shamlian) [reply docs/webserver/rmw-profile-override.md:10]: They are already on that page, no?

File: docs/api/hazards.md (hunk 3 above)
Comment (shamlian) [reply docs/api/hazards.md:171]: Addeded in collapsible admonitions with some examples from `ros2 topic echo`

**Pull request merged**